packages/push/src/instruction/common/pop.rs

```diff
@@ -0,0 +1,140 @@
+/// | ------------- | ------------- | ------------- |
+/// | exists  | ✅ | The top value is removed |
+/// | missing | [❗…](crate::push_vm::stack::StackError::Underflow) | State is unchanged |
+///
+/// # Errors
+///
+/// The only possible error here is
+/// [`StackError::Underflow`](crate::push_vm::stack::StackError::Underflow)
+/// and the instruction will then be ignored.
+#[derive(Debug, Copy, Clone, Default, Eq, PartialEq)]
+pub struct Pop<T> {
+    _p: PhantomData<T>,
+}
+
+impl<T> Pop<T> {
+    pub const fn new() -> Self {
+        Self { _p: PhantomData }
+    }
+}
+
+impl<S, T> Instruction<S> for Pop<T>
+where
+    S: Clone + HasStack<T>,
+    T: Clone,
+{
+    type Error = PushInstructionError;
+
+    fn perform(&self, mut state: S) -> InstructionResult<S, Self::Error> {
+        match state.stack_mut::<T>().pop() {
+            Ok(_) => Ok(state),
+            // If `pop()`` fails it's because the stack was empty and we
+            // want to return a recoverable error so the `pop` instruction
+            // can be ignored.
+            Err(error) => Err(Error::recoverable(state, error)),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use proptest::{collection::vec, prelude::*};
+    use test_strategy::proptest;
+
+    use super::*;
+    use crate::push_vm::{push_state::PushState, stack::StackError};
+
+    #[test]
+    fn pop_empty_stack() {
+        let state = PushState::builder()
+            .with_max_stack_size(0)
+            .with_no_program()
+            .build();
+        let pop = Pop::<i64>::default();
+        let result = pop.perform(state).unwrap_err();
+        assert!(result.is_recoverable());
+        assert_eq!(
+            result.error(),
+            &PushInstructionError::StackError(StackError::Underflow {
+                num_requested: 1,
+                num_present: 0
+            })
+        );
+    }
+
+    #[test]
+    fn pop_non_empty_stack() {
+        let state = PushState::builder()
+            .with_max_stack_size(1)
+            .with_int_values([17])
+            .unwrap()
+            .with_no_program()
+            .build();
+        let pop = Pop::<i64>::default();
+        let result = pop.perform(state).unwrap();
+        assert_eq!(result.stack::<i64>().size(), 0);
+    }
+
+    // Test that `Pop` works on stacks of 1..100 elements of type `i64`.
+    #[proptest]
+    fn pop_stacks_with_multiple_items(#[strategy(vec(any::<i64>(), 1..100))] values: Vec<i64>) {
+        let state = PushState::builder()
+            .with_max_stack_size(100)
+            .with_int_values(values.clone().into_iter().rev())
+            .unwrap()
+            .with_no_program()
+            .build();
+        let pop = Pop::<i64>::default();
+        let result = pop.perform(state).unwrap();
+        #[expect(
+            clippy::arithmetic_side_effects,
+            reason = "The length of `values` is less than 100, so adding 1 is guaranteed to be \
+                      safe."
+        )]
```

Comment:
this safety comment doesn't make any sense in the context of the arithmetic operation we actually perform